### ISSUE TITLE
DefaultClaimsService.GetIdentityTokenClaimsAsync uses wrong Resource parameter for ProfileData

### ISSUE THREAD

**Issue by kgaute** [IdentityServer/IdentityServer4#4900]
Version: 4.1.0

```
public virtual async Task<IEnumerable<Claim>> GetIdentityTokenClaimsAsync
(
	ClaimsPrincipal subject, 
	ResourceValidationResult resources,  // contains Resources validated by   
 //  TokenResponseGenerator.ProcessAuthorizationCodeRequestAsync()
	bool includeAllIdentityClaims, 
	ValidatedRequest request
)
```

```
var context = new ProfileDataRequestContext(
    subject,
    request.Client,
    IdentityServerConstants.ProfileDataCallers.ClaimsProviderIdentityToken,
    additionalClaimTypes)
{
    RequestedResources = request.ValidatedResources, // is empty
    ValidatedRequest = request
};	
```

As the title says the `ProfileDataRequestContext` created for the `ProfileService` (IdentityToken) will never contain 
a valid `RequestedResources` object. The error appears to stem from the `DefaultClaimsService.GetIdentityTokenClaimsAsync(...)` where the `RequestedResources` is set as the `request.ValidatedResources`, however this property (as far as I can tell) is never touched. I'd assume the intended property to set is the parameter to GetIdentityTokenClaimsAsync; `ResourceValidationResult` which contains the expected resources. 

I'd be happy to submit a PR if my assumption is correct. 

### Issue / Steps to reproduce the problem
Create a ProfileService where the IdentityToken's claims are populated based on IdentityResources, not directly from Claims.

**Comment by brockallen** [comment]
> As the title says the ProfileDataRequestContext created for the ProfileService (IdentityToken) will never contain
a valid RequestedResources object. 

Can you explain more? I don't understand what you mean. In what scenario? Which response_type and what scopes?

**Comment by IndexOverflow** [comment]
(Sorry about the account mix-up)

The scenario is when going to the ProfileService for fetching claims for the id_token using an auth code (response_type=code). I believe any "identity resource bound" scope, but at least "profile". 

As I tried to show in my example the `resources` parameter contains the expected IdentityResources, but the ProfileDataRequestContext is populated with the parameter from request.ValidatedResources which isn't populated. 
The fix was simply then,

```
                var context = new ProfileDataRequestContext(
                    subject,
                    request.Client,
                    IdentityServerConstants.ProfileDataCallers.ClaimsProviderIdentityToken,
                    additionalClaimTypes)
                {
                    RequestedResources = resources,
                    ValidatedRequest = request
                };
```

**Comment by brockallen** [comment]
For code flow, the identity token (by default) never has claims. Those are normally obtained via the userinfo endpoint. But if you *do* want the claims in the id_token, then on the Client enable the `AlwaysIncludeUserClaimsInIdToken` setting:

https://identityserver4.readthedocs.io/en/latest/reference/client.html#token

**Comment by IndexOverflow** [comment]
I'm aware, this is with that flag set. I did however notice that claims found  in the resource are propagated through as expected (additionalClaimTypes). But I'm still not understanding why the ResourceValidationResult is dismissed in favor of the request.ValidatedResources. It may be a faulty design on my end, but I've elected to populate the tokens & userprofile based on the requested resources rather than looking at the claims (as there are many, dynamic claimTypes in play).
